sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -431,7 +431,8 @@ case class AdaptiveSparkPlanExec(
               // Check the `stageCache` again for reuse. If a match is found, ditch the new stage
               // and reuse the existing stage found in the `stageCache`, otherwise update the
               // `stageCache` with the new stage.
-              val queryStage = context.stageCache.getOrElseUpdate(e.canonicalized, newStage)
+              val queryStage = context.stageCache.getOrElseUpdate(
+                newStage.plan.canonicalized, newStage)
```

Comment:
This breaks DPP. I'm not sure exactly why, but the key of stage cache was plan without applying `queryStageOptimizerRules`, now it's different. This probably breaks DPP.

Reply:
One idea is to introduce a new phase: `preStageCreationRules`. We can put `ApplyColumnarRulesAndInsertTransitions` and run it before reusing exchanges.

Reply:
Thanks for the suggestions. I am working on this now and will try this out.

Reply:
thanks for pointing it out. That feels odd to me that it breaks the DPP. I did also notice seems more changes to the way that works in flight: https://github.com/apache/spark/pull/31756/files

Reply:
If we merge https://github.com/apache/spark/pull/31756 and then rebase this PR the DPP tests pass so I think this is the best path forward.

Reply:
#31756 was merged @andygrove  can you up merge and retest to make sure this works?